### DISCUSSION TITLE
gg.jte.TemplateNotFoundException: Failed to load index.jte

### DISCUSSION THREAD

**Discussion by kenduraghav** [casid/jte#391] [Q&A]
I am just started to explore JTE. I have been trying to work with Servlet as backend and rendering the output in JTE.  I am using maven based project. 

```
@WebServlet("/hello-jte")
public class HelloServlet extends HttpServlet {
	private static final long serialVersionUID = 1L;

	TemplateEngine templateEngine;

        public HelloServlet() {
		  super();
		  templateEngine = TemplateEngine.createPrecompiled(Path.of("jte-classes"), ContentType.Html);
		  ;
		  // templateEngine = TemplateEngine.createPrecompiled(ContentType.Html);;
	  }

	protected void doGet(HttpServletRequest request, HttpServletResponse response)
			throws ServletException, IOException {
		TemplateOutput output = new StringOutput();
		String name = "John Doe";
		templateEngine.render("index.jte", name, output);
		System.out.println(output);

		response.setContentType("text/html");
	}

	protected void doPost(HttpServletRequest request, HttpServletResponse response)
			throws ServletException, IOException {
	}

}
``` 

However, I am getting the below error. 

![image](https://github.com/user-attachments/assets/d211c4b1-7d62-4112-9fd6-3c134bc4bedb)

Here is my Github [repository](https://github.com/kenduraghav/jte-demo.git)

My Query Is this workable using servlets.? Because i have some legacy application which i wanted to migrate to jte  if this works with servlet. 

Kindly  advise me.

**Comment by casid** [comment]
Hi @kenduraghav,

this should work within a servlet. I think the problem is that you use `createPrecompiled`, but haven't precompiled the templates.

During development it is recommended to have hot-reloaded templates that are compiled on the fly, and to precompile the templates for production builds in the build process. You can find some docs about it here: https://jte.gg/pre-compiling/

**Comment by kenduraghav** [comment]
Hi Casid/Jte

Thanks for the reply.

I have tried both incase of development , it’s  trying to load from my
eclipse installation location but not under the project src/main/jte

Here I am sharing the screenshot and I hope you might be busy in your work
and if you got a time to look my repository it would be helpful

I am attaching my repo link as well . Kindly visit at your convenience if
possible

https://github.com/kenduraghav/jte-demo/blob/main/error-log.md





Thanks & Regards,
Raghav

**Comment by kenduraghav** [comment]
<p> Here is the error while using code resolver syntax for the development</p>
<code>
CodeResolver codeResolver = new DirectoryCodeResolver(Path.of("src/main/jte"));<br>
templateEngine = TemplateEngine.create(codeResolver, ContentType.Html);
</code>

![error1](https://github.com/user-attachments/assets/742036c0-7c89-4788-8861-94036fa8c274)

**Comment by fride** [comment]
Same problem here. There is a file `src/main/index.jte` but it fails with `gg.jte.TemplateNotFoundException: src/main/jte/index.jte not found`.

For context I'm using Javalin and I'm running the app from Idea.

**Reply by fride** [reply]
Ok, shame on me! :D It works when I use the correct file name.

**Comment by sysmat** [comment]
- I have the same problem, I want to deploy war to app server `Caused by: gg.jte.TemplateNotFoundException: Failed to load hello.jte`

1. I use maven and plugin for pre compiled
2. I want  TemplateEngine.createPrecompiled
3. jte file in resource/jte/a resource/jte/b

What is connection plugin targetDirectory & createPrecompiled class dir? 
- When I build with maven I see `JtehelloGenerated.class` it in  ` \WEB-INF\classes\gg\jte\generated\precompiled`
- I set `Path targetDirectory = Paths.get("gg");` but still Failed to load hello.jte

```
<plugin>
                <groupId>gg.jte</groupId>
                <artifactId>jte-maven-plugin</artifactId>
                <version>${jte.version}</version>
                <configuration>
                    <sourceDirectory>${project.basedir}/src/main/resources/jte</sourceDirectory>
                    <targetDirectory>${project.build.directory}/jte-classes</targetDirectory>
                    <contentType>Plain</contentType>
                </configuration>
                <executions>
                    <execution>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>generate</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

```
Path targetDirectory = Paths.get("gg"); // This is the directory where compiled templates are located.
TemplateEngine.createPrecompiled(targetDirectory, ContentType.Plain);
```

**Reply by sysmat** [reply]
- even if I use DirectoryCodeResolver("jte") and `.render("hello.jte", username, output);`  in `WEB-INF\classes\jte\hello.jte` and still `TemplateNotFoundException: hello.jte not found` 

- nice lib, but if I can figure it out how to render template in war it is not for me

**Reply by casid** [reply]
Hi @sysmat, I think you probably want a self contained war file, that contains the entire application including the jte templates, right? Have you tried this guide? https://jte.gg/pre-compiling/#using-the-application-class-loader

**Reply by sysmat** [reply]
Hi @casid thx, you make my day. I didn't find in docs

**Reply by casid** [reply]
No worries, glad it is working now!
